### PR TITLE
Support preopening NeXus files

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is not strictly required by anything right now, but I needed it for some testing, so a cleaned it up for the future.

Changes:
- Be more relaxed about passing definitions for open groups.
- Add flag to optionally open NeXus files. This may become more relevant in the future when loading even more different parts and/or chunks from files.

### CODE REVIEW

File: src/ess/reduce/nexus/generic_workflow.py [hunk 1]
```diff
@@ -17,15 +17,15 @@
...
 
 
-def file_path_to_file_spec(filename: gt.Filename[RunType]) -> gt.NeXusFileSpec[RunType]:
-    return gt.NeXusFileSpec[RunType](filename)
+def file_path_to_file_spec(
+    filename: gt.Filename[RunType], preopen: PreopenNeXusFile
+) -> gt.NeXusFileSpec[RunType]:
+    return gt.NeXusFileSpec[RunType](
+        snx.File(filename, definitions=workflow.definitions) if preopen else filename
+    )
```
Comment (jl-wynen) [inline src/ess/reduce/nexus/generic_workflow.py:28]: I don't like this. Normally, we would open the file and pass the handle to a loader. I would expect the same from a workflow. That is, I would want to do this:
```python
handle = snx.File(...)
params[gt.Filename[SampleRun]] = handle
```

I thought the duck-typing approach would allow this?

File: src/ess/reduce/nexus/workflow.py [hunk 2]
```diff
@@ -475,6 +471,11 @@ def _add_variances(da: sc.DataArray) -> sc.DataArray:
     return out
 
 
+definitions = snx.base_definitions()
+definitions["NXdetector"] = _StrippedDetector
+definitions["NXmonitor"] = _StrippedMonitor
```
Comment (jl-wynen) [inline src/ess/reduce/nexus/workflow.py:476]: Can you use a function or an immutable dict instead? I don't like mutable global (public) variables.

File: src/ess/reduce/nexus/generic_workflow.py [hunk 3]
```diff
@@ -220,24 +220,34 @@ def assemble_monitor_data(
     assemble_detector_data,
 )
 
+_default_params = {
+    PulseSelection: PulseSelection(()),
```
Comment (jl-wynen) [inline src/ess/reduce/nexus/generic_workflow.py:224]: Why is this needed now? It seems unrelated to the other changes.

Comment (SimonHeybrock) [reply src/ess/reduce/nexus/generic_workflow.py:224]: Avoids 3x code duplication?

File: src/ess/reduce/nexus/generic_workflow.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/ess/reduce/nexus/generic_workflow.py:28]: Well, that would not pass mypy, since `Filename` cannot be a file handle?

Comment (jl-wynen) [reply src/ess/reduce/nexus/generic_workflow.py:28]: Isn't this the whole point of not checking the types of parameters any more?

You are right that, in principle, it shouldn't pass. But it does pass.

File: src/ess/reduce/nexus/generic_workflow.py (hunk 3 above)
Comment (SimonHeybrock) [reply src/ess/reduce/nexus/generic_workflow.py:224]: You are right, that was a bad rebase, fixing.

File: src/ess/reduce/nexus/generic_workflow.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/ess/reduce/nexus/generic_workflow.py:28]: I am not sure I understand what you are saying. There appear to be 3 different things you mention: Having a provider open a file, mypy, and not checking types of parameters in Sciline.

They all seem kind of distinct. We should have a chat so you can explain!

I would like to control the behavior via a parameter, since having to change a ton of user code in a very complicated manner does not seem feasible or remotely user friendly.

**Comment by jl-wynen** [comment]
> @jl-wynen is this related to [scipp/scippnexus#238](https://github.com/scipp/scippnexus/pull/238) ?

Yes. But ESSreduce has its own mechanism for this. See https://github.com/scipp/essreduce/blob/dd87cd7264f0187dd888ce6afcd7096df2ca5391/src/ess/reduce/nexus/_nexus_loader.py#L65

**Comment by SimonHeybrock** [comment]
As mentioned in the discussion related to (not) locking (#99) earlier, I think doing this via an injected provider seems like a decent option to me. The question is whether having what we have in this PR as a default is reasonable, or if should be something else.

As this sees little use for now, we can also consider to go with it, then iterate.

Please review again under those considerations!

File: src/ess/reduce/nexus/workflow.py (hunk 2 above)
Comment (jl-wynen) [reply src/ess/reduce/nexus/workflow.py:509]: This comment still applies.

**Comment by jl-wynen** [comment]
> As mentioned in the discussion related to (not) locking (https://github.com/scipp/essreduce/issues/99) earlier, I think doing this via an injected provider seems like a decent option to me.

What do you mean by 'injected provider'? I don't see one in this PR. So are you saying that we should do it differently?

**Comment by SimonHeybrock** [comment]
> > As mentioned in the discussion related to (not) locking (#99) earlier, I think doing this via an injected provider seems like a decent option to me.
> 
> What do you mean by 'injected provider'? I don't see one in this PR. So are you saying that we should do it differently?

This PR (re)defines `file_path_to_file_spec`. What I am saying is that if a user requires special features such as non-locking or SWMR mode they can insert their own into the workflow, kicking out the pre-defined one.

**Review by jl-wynen — APPROVED** [review]
